packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -299,6 +311,111 @@ export const useInvokeAPI: EnsembleActionHook<InvokeAPIAction> = (action) => {
   return invokeApi;
 };
 
+export const useConnectSocket: EnsembleActionHook<ConnectSocketAction> = (
+  action,
+) => {
+  const screenData = useScreenData();
+  const [isOpen, setIsOpen] = useState<boolean>();
+
+  const socket = useMemo(
+    () => screenData.sockets?.find((model) => model.name === action?.name),
+    [action, screenData],
+  );
+
+  const onSocketConnectAction = useEnsembleAction(socket?.onSuccess);
+  const onMessageReceiveAction = useEnsembleAction(socket?.onReceive);
+  const onSocketDisconnectAction = useEnsembleAction(socket?.onDisconnect);
+
+  const connectSocket = useMemo(() => {
+    if (!socket) {
+      return;
+    }
+
+    const callback = (): void => {
+      setIsOpen(false);
+    };
+    return { callback };
+  }, [socket]);
+
+  useEffect(() => {
+    if (!socket || isOpen !== false) {
+      return;
+    }
+
+    handleConnectSocket(
+      screenData,
+      socket.name,
+      onSocketConnectAction,
+      onMessageReceiveAction,
+      onSocketDisconnectAction,
+    );
+    setIsOpen(true);
+  }, [
+    socket,
+    isOpen,
+    screenData,
+    onSocketConnectAction,
+    onMessageReceiveAction,
+    onSocketDisconnectAction,
+  ]);
+
+  return connectSocket;
+};
+
+export const useMessageSocket: EnsembleActionHook<SendSocketMessageAction> = (
+  action,
+) => {
+  const screenData = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+  const [context, setContext] = useState<{ [key: string]: unknown }>();
+  const evaluatedInputs = useEvaluate(action?.message, { context });
+
+  const sendSocketMessage = useMemo(() => {
+    const callback = (args: unknown): void => {
+      setIsComplete(false);
+      setContext(args as { [key: string]: unknown });
+    };
+    return { callback };
+  }, []);
+
+  useEffect(() => {
+    if (!action || isComplete !== false) {
+      return;
+    }
+
+    // send socket message
+    handleMessageSocket(screenData, action.name, evaluatedInputs);
+    setIsComplete(true);
+  }, [screenData, action, evaluatedInputs, isComplete]);
+
+  return sendSocketMessage;
+};
+
+export const useDisconnectSocket: EnsembleActionHook<DisconnectSocketAction> = (
+  action,
+) => {
+  const screenData = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+
+  const disconnectSocket = useMemo(() => {
+    const callback = (): void => {
+      setIsComplete(false);
+    };
+    return { callback };
+  }, []);
+
+  useEffect(() => {
+    if (!action || isComplete !== false) {
+      return;
+    }
+
+    handleDisconnectSocket(screenData, action.name);
+    setIsComplete(true);
+  }, [screenData, action, isComplete]);
```

Comment:
Don't need to track completion state here just return a callback with the handleDisconnect fn.